### PR TITLE
Implement TargetedLocationData

### PR DESCRIPTION
**Pull request by kamcio96**
@gabizou 
```
An {@link DataManipulator} handling the supposed targetted
{@link Location}. Usually for the case of {@link EnderPearl}s, the
targeted {@link Location} is where the {@link EnderPearl} will
move towards until it's expiration time. In the case of
{@link ItemStack}s of type {@link ItemTypes#COMPASS}, the targeted
{@link Location} is where the compass will point towards.
```
1. EntityLiving with target location.
2. EnderPearl mean EntityEnderEye?
3. Player with compass target.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
+package org.spongepowered.common.data.processor.data;
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java]: Missing license header.

File: src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public Optional<TargetedLocationData> from(DataHolder dataHolder) {
+        if(dataHolder instanceof EntityPlayer) {
+            EntityPlayer player = (EntityPlayer) dataHolder;
+            IMixinEntityPlayer mixinPlayer = (IMixinEntityPlayer) dataHolder;
+            BlockPos pos = mixinPlayer.getCompassLocation();
+            if(pos == null) pos = player.getEntityWorld().getSpawnPoint();
+            final TargetedLocationData data = new SpongeTargetedLocationData(new Location<>((World) player.getEntityWorld(), pos.getX(), pos.getY(), pos.getZ()));
+            return Optional.of(data);
+        } else if(dataHolder instanceof EntityLiving) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java]: Pretty sure I didn't intend this to do anything with the navigator.

**Comment by kamcio96** [comment]
Where I have to implement set methods? 
```
org.spongepowered.common.data.processor.value.entity.EntityTargetedLocationValueProcessor
```
?

File: src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import org.spongepowered.common.data.manipulator.mutable.SpongeTargetedLocationData;
+import org.spongepowered.common.data.processor.common.AbstractSpongeDataProcessor;
+import org.spongepowered.common.data.util.DataUtil;
+import org.spongepowered.common.interfaces.IMixinEntityPlayer;
+
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class TargetedLocationDataProcessor extends AbstractSpongeDataProcessor<TargetedLocationData, ImmutableTargetedLocationData> {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java]: Use `AbstractEntitySingleDataProcessor`, it implements a bunch of methods already

File: src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    @Override
+    public Optional<TargetedLocationData> createFrom(DataHolder dataHolder) {
+        if (dataHolder instanceof EntityPlayer) {
+            return from(dataHolder);
+        } else if (dataHolder instanceof EntityLiving) {
+            if (((EntityLiving) dataHolder).getNavigator().getPath() != null) {
+                return from(dataHolder);
+            } else {
+                return Optional.empty();
```
Comment (JBYoshi) [inline src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java]: Should be `return Optional.of(new SpongeTargetedLocationData());`

**Comment by kamcio96** [comment]
@gabizou EnderPearl in javadoc means EntityEnderEye?

**Comment by JBYoshi** [comment]
Yep.

**Comment by Aaron1011** [comment]
A couple things:

* I'm not sure if this should represent the pathfinder data. That would probably just be a method (not data), since it's not serialized.
* Can you open a PR against SpongeAPI's `bleeding` branch? There's a few things that probably should be change:
 * `TargetedLocationData` is applicable to `Player`s, not `ItemStack`s.
 * The value and corresponding `Key` should be a `Vector3d`, not a `Location`. Both the player and the `EyeOfEnder` only receive a position, which they use with the world they exist in.

File: src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import org.spongepowered.common.SpongeImpl;
+import org.spongepowered.common.data.processor.common.AbstractEntitySingleDataProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.interfaces.IMixinEntityPlayer;
+import org.spongepowered.common.interfaces.entity.IMixinEntityEnderEye;
+import org.spongepowered.common.util.VecHelper;
+
+import java.util.Optional;
+
+public class TargetedLocationDataProcessor extends AbstractEntitySingleDataProcessor<net.minecraft.entity.Entity, Location<World>, Value<Location<World>>, TargetedLocationData, ImmutableTargetedLocationData> {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java]: Formatting throughout this file does not follow the code style.

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayer.java [hunk 6]
```diff
@@ -135,4 +138,24 @@ public boolean affectsSpawning() {
     public void setAffectsSpawning(boolean affectsSpawning) {
         this.affectsSpawning = affectsSpawning;
     }
+
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayer.java]: Formatting does not follow the code style.

**Comment by kamcio96** [comment]
@Aaron1011 So what I have to change?

**Comment by Aaron1011** [comment]
@kamcio96: I would remove the logic dealing with the pathfinder target, and make the API changes I suggested.

File: src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof EntityEnderEye || dataHolder instanceof EntityLiving;
+    }
+
+    @Override
+    protected TargetedLocationData createManipulator() {
+        //return new SpongeTargetedLocationData(new Location<World>()); what there?
+        return null;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/TargetedLocationDataProcessor.java]: Should likely default to the overworld's spawn point?

**Comment by gabizou** [comment]
@kamcio96 Please fix your formatting, we provide the code style for both IntelliJ and Eclipse in the API repository. Likewise, you aren't registering your processor correctly, if it's supposed to be a "dual" processor, you need to rebase this off at least the commit adding the processor cleanup, or to current SpongePowered's master. Finally, I need some test plugin code to actually test this implementation.

**Comment by kamcio96** [comment]
@gabizou sorry I'm little busy now, have I change api too?

**Comment by kamcio96** [comment]
hmmm, could you update api please?
i don't understand what you really want

**Comment by gabizou** [comment]
@kamcio96 You need to rebase this branch on top of master and resolve any conflicts that exist.

**Comment by kamcio96** [comment]
@gabizou I'm really busy now, so i can't focus on that. Feel free to close and make your commit.

**Comment by kashike** [comment]
I'll finish this in a new PR.

**Comment by kamcio96** [comment]
@kashike nice kitty :wink:

**Pull request closed without merging**